docs-test-gen/src/main.rs

```diff
@@ -61,86 +61,90 @@ struct CodeBlock {
+            if let Event::Text(ref text) = event {
+                code.push_str(text);
+            }
+        }
 
-        let mut language = Language::Rust;
-        let mut template = String::new();
-        let mut agg = String::new();
-        let mut blocks = Vec::new();
+        match event {
+            Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(fence))) => {
+                language = match find_language(&fence) {
+                    Some(found_lang) => found_lang,
+                    None => continue,
+                };
 
-        for event in parser {
-            if is_aggregating {
-                if let Event::Text(ref text) = event {
-                    agg.push_str(&text);
+                if let Some(template_name) = get_template_name(&fence) {
+                    is_aggregating = true;
+                    template = template_name.to_string();
                 }
             }
+            Event::End(TagEnd::CodeBlock) if is_aggregating => {
+                is_aggregating = false;
 
-            match event {
-                Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(fence))) => {
-                    if is_rusty(&fence) {
-                        language = Language::Rust;
-                    } else if is_goey(&fence) {
-                        language = Language::Go;
-                    } else {
-                        continue;
-                    }
-
-                    if let Some(template_name) = get_template_name(&fence) {
-                        is_aggregating = true;
-                        template = template_name.to_string();
-                    }
-                }
-                Event::End(TagEnd::CodeBlock) if is_aggregating => {
-                    is_aggregating = false;
-
-                    blocks.push(CodeBlock {
-                        template: template.clone(),
-                        code: agg.clone(),
-                        language,
-                    });
+                blocks.push(CodeBlock {
+                    template: template.clone(),
+                    code: code.clone(),
+                    language,
+                });
 
-                    template.clear();
-                    agg.clear();
-                }
-                _ => (),
+                template.clear();
+                code.clear();
             }
+            _ => (),
         }
+    }
 
-        if blocks.is_empty() {
-            continue;
-        }
+    for (idx, block) in blocks.iter().enumerate() {
+        let mdx_filename = path.file_name().unwrap().to_str().ok_or_else(|| {
+            anyhow!(
+                "filename of path \"{}\" isn't valid utf-8. (really?! in the 21st century?)",
+                path.display()
+            )
+        })?;
+
+        let filename = format!(
+            "{}_{}_{}.{}",
+            mdx_filename.replace('.', "_"),
+            block.template,
+            idx,
+            block.language.file_ext(),
+        );
```

Comment:
One thing I noticed while running this locally just now: If two files have the same name and template (e.g. one getting-started.md in multitest, one in ibc), this could produce collisions that silently only test one. It's not very likely, but would be good to avoid it since we won't notice it otherwise.
Maybe we can include the relative path after `src` or so?

Reply:
Yeah, might be good to do that. The initial design was limited by the idea of "I don't want files with potentially 200 characters committed into the repo".  
But now that we went down the route of generating it just before we run the tests, we can do that.